src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/bytecodes/ReturnBytecodes.java

```diff
@@ -69,7 +76,37 @@ protected final Object doNonLocalReturn(final VirtualFrame frame) {
         @Specialization(guards = {"isCompiledBlockObject(code)"})
         protected final Object doClosureReturn(final VirtualFrame frame) {
             // Target is sender of closure's home context.
-            throw new NonLocalReturn(getReturnValue(frame), FrameAccess.getClosure(frame).getHomeContext().getFrameSender());
+            final ContextObject homeContext = FrameAccess.getClosure(frame).getHomeContext();
+            final ContextObject[] firstUsableContext = new ContextObject[1];
+            final boolean homeContextNotOnTheStack = null == Truffle.getRuntime().iterateFrames(frameInstance -> {
```

Comment:
Thanks for the PR! Just briefly looked at this and saw that you use `iterateFrames()` for all non-local returns. That's probably going to be *very* slow. Instead, we may want to add a flag to `ContextObject` that indicates if the context is active on the stack or not. We've been doing something similar in RSqueak (see [here](https://github.com/hpi-swa/RSqueak/blob/7d9b6649efeb1d1fff9f2ecf4084d3dd18082022/rsqueakvm/storage_contexts.py#L16-L32)).

Reply:
I had a bug that I had to fix, so I re-pushed.
Now it passes the CI tests (it does show a Checkstyle error, but it's not something I modified or introduced)
I also added some specializations to try to avoid as much as possible materialization and the frame iteration.
I am afraid a flag on context would not help. What we are looking for with this iteration, and which is the "common" case for a cannotReturn situation, is a context that is valid and active on a stack, just not the right one (it's in a different process, see the testcases)

Reply:
What I think would be interesting to pursue, and I am sure that would help performance in general, not just by reducing these iterations, is targeted and (more?) aggressive inlining of the methods around closures, until the closures disappear - then the non-local returns disappear as well.

Reply:
> But I am afraid that a flag on the context would not help. What we are looking for within this iteration, and which is the "common" case for a cannotReturn, is a context that is valid and active on a stack, just not the right one (e.g. it's in a different process, see the testcases)

By a flag on the context that indicates whether it is active or not, I meant active on the Truffle stack. Only one process can be active on the Truffle stack, so this should work, right?

> What I think would be interesting to pursue, and I am sure that would help performance in general, not just by reducing these iterations, is targeted and (more?) aggressive inlining of the methods around closures, until the closures disappear - then the non-local returns disappear as well.

The Graal compiler inlines closures already. However, I am not sure how well it can figure out the stack walking magic (`iterateFrames`). AFAIK, `iterateFrames` should really only be used the first time on the slow path, and afterwards, there should be an alternative version that can be compiled as part of the fast path.

Reply:
You are probably right, the flag should work, at least for "normal" code. I am not sure how e.g. copyStack-based frameworks like Seaside are (or at least used to be) implemented - anyway, such code would probably be a good testcase.

Reply:
Implementing the flag should be relatively straightforward (compared to other approaches), shall we give it a try and take it from there? BTW: we ran the Seaside test suite at ESUG and were surprised that only a couple of Context-related tests are not passing at the moment. Not sure whether you've seen [this](https://github.com/marschall/seaside-servlet-squeak). :)

Reply:
Very nice, I had not seen it, thanks!
I'll try to implement the flag